File: problems/设计/225.用队列实现栈.cpp
#include <iostream>
#include <queue>
using namespace std;

class MyStack {
    queue<int> q;
public:
    void push(int x) {
        q.push(x);
        for (int i = 0; i < (int)q.size() - 1; i++) {
            q.push(q.front());
            q.pop();
        }
    }
    int pop() {
        int v = q.front();
        q.pop();
        return v;
    }
    int top() {
        return q.front();
    }
    bool empty() {
        return q.empty();
    }
};

int main() {
    // 测试用例
    MyStack st;
    st.push(1);
    st.push(2);
    cout << st.top() << endl;   // 预期: 2
    cout << st.pop() << endl;  // 预期: 2
    cout << boolalpha << st.empty() << endl;  // 预期: false
    st.push(3);
    cout << st.pop() << endl;  // 预期: 3
    cout << st.pop() << endl;  // 预期: 1
    cout << st.empty() << endl;  // 预期: true
    return 0;
}

File: problems/设计/225.用队列实现栈.py
#!/usr/bin/env python3
"""
225. 用队列实现栈
"""
from collections import deque


class MyStack:
    def __init__(self):
        self.q = deque()

    def push(self, x: int) -> None:
        self.q.append(x)
        for _ in range(len(self.q) - 1):
            self.q.append(self.q.popleft())

    def pop(self) -> int:
        return self.q.popleft()

    def top(self) -> int:
        return self.q[0]

    def empty(self) -> bool:
        return not self.q


if __name__ == "__main__":
    # 测试用例
    stack = MyStack()
    stack.push(1)
    stack.push(2)
    print(stack.top())   # 预期: 2
    print(stack.pop())   # 预期: 2
    print(stack.empty())  # 预期: False
    stack.push(3)
    print(stack.pop())   # 预期: 3
    print(stack.pop())   # 预期: 1
    print(stack.empty()) # 预期: True
